MP JWT fixes

Ladicek opened the pull request:
- [THORN-2142](https://issues.jboss.org/browse/THORN-2142)
- [THORN-2143](https://issues.jboss.org/browse/THORN-2143)
- [THORN-2144](https://issues.jboss.org/browse/THORN-2144)

Tests for THORN-2142 and THORN-2143 are already present (and currently failing) in https://github.com/rhoar-qe/thorntail-test-suite. Test for THORN-2144 added there as well.

- [x] Have you followed the guidelines in our [Contributing](http://wildfly-swarm.io/community/contributing/) document?
- [x] [v2] Have you created a [JIRA](https://issues.jboss.org/browse/THORN) and used it in the commit message?
- [ ] [v4] Have you created a [GitHub Issue](https://github.com/thorntail/thorntail/issues) and used it in the commit message?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/thorntail/thorntail/pulls) for the same issue?
- [ ] Have you built the project locally prior to submission with `mvn clean install`?

Ladicek committed:
THORN-2143: MP JWT: using * in @RolesAllowed doesn't work

Motivation
----------
After the recent refactoring of MP JWT authorization, annotating
a resource method with `@RolesAllowed("*")` no longer works.
This is inconsistent with previous behavior and also with
how security annotations behave by default.

Modifications
-------------
If one of the allowed roles is `*`, then access is allowed
for all logged-in users.

Result
------
Using `@RolesAllowed("*")` works as expected again.

Ladicek committed:
THORN-2142: MP JWT: resource methods without security annotations are always allowed

Motivation
----------
After the recent refactoring of MP JWT authorization, JAX-RS resource
methods that don't have a security annotation (and the class doesn't
have a security annotation either) are always allowed.

This is inconsistent with previous behavior and also with how security
annotations behave by default. If the class contains other methods that
do have security annotations, then methods without security annotations
should behave as if they were annotated `@DenyAll`. This is also
documented in the fraction reference docs; see
`swarm.microprofile.jwt.default-missing-method-permissions-deny-access`.

Modifications
-------------
The JAX-RS dynamic feature that implements role-based access control
now automatically denies access to a resource method if:

- it doesn't have a security annotation, and
- the class doesn't have a security annotation, and
- some other resource method in the class has a security annotation.

The `default-missing-method-permissions-deny-access` configuration is
also respected. It defaults to `true`, which enables the behavior
described above, but if set to `false`, resource methods without
security annotations are allowed.

Result
------
Role-based access control to JAX-RS resource methods per MP JWT
works as expected again.

Ladicek committed:
THORN-2144: MP JWT: configuring token expiration grace period via project-defaults.yml doesn't work

Motivation
----------
After the recent refactoring of MP JWT authorization, configuring
the MP JWT token expiration grace period via `project-defaults.yml`
using `swarm.microprofile.jwt.token.exp-grace-period` no longer works.
This is inconsistent with previous behavior.

Modifications
-------------
Propagate fraction configuration to the deployment, so that
the `JWTAuthMechanismFactory` can read it and apply.

Result
------
Configuring MP JWT token expiration grace period works
as expected again.

Ladicek commented on fractions/microprofile/microprofile-jwt/src/main/java/org/wildfly/swarm/microprofile/jwtauth/MpJwtFilterRegistrar.java:
```diff
@@ -92,4 +100,21 @@ private Annotation getAnnotation(Annotation[] declaredAnnotations,
     }
+
+    private boolean hasSecurityAnnotations(ResourceInfo resource) {
+        // TODO check all methods of the class, or just resource methods? public/private/etc.? inherited?
```
@mkouba I recall you were the one to add `swarm.microprofile.jwt.default-missing-method-permissions-deny-access`. Could you advise here? I think our docs say that only resource methods are considered here (http://docs.wildfly-swarm.io/2.1.0.Final/#_microprofile_jwt_rbac_auth), but not sure about e.g. inheritance. If only JAX-RS's `ResourceInfo` provided a list of all resource methods :-)

Ladicek replied:
Ah Martin is on PTO this week. Hmm. Anyone? :-)

sberyozkin replied:
Does the spec require that only security annotated methods can be invoked ? Role based access is an extra restriction...I.e, one service can be happy enough with the token be validated (client - authenticated only)

Ladicek replied:
Implicit `@DenyAll` for all methods in a class that don't have a security annotation yet some other method does mirrors the behavior we have for EJBs. The configuration key and its default value does the same. Not sure if that answers your question :-)

sberyozkin replied:
Hi Ladislav, I was not sure, if the token can only be used as the authorization token, which what would happen with your change. I've looked at the article about MP JWT, my understanding it should be also usable as only the authentication token, ie, if it is valid hence the client is authenticated and the service is free to access the claims without having to put RolesAllowed, etc...
thanks

Ladicek replied:
I originally assumed the same -- see https://issues.jboss.org/browse/THORN-1954. In this PR, I'm just returning to our own previous behavior.

sberyozkin replied:
Right, if adding something like @RolesAllowed("*") will still work then it is fine, even if it might be a bit of an extra work.

Ladicek replied:
`@RolesAllowed("*")` is another thing I'm restoring in this PR :-)

Ladicek commented:
Hmm I'll take a look at the failing test.

EDIT: yea, that's very simple, the test expects that a method without security annotation is always accessible, which is a bug. Test is very easy to fix.

Ladicek committed:
THORN-2110 followup: improve the principal leak test

Motivation
----------
The `PrincipalLeakTest` assumed that the number of worker threads
is always 64. That, however, isn't true -- it's actually derived
from the number of currently available CPU cores.

It's much better to explicitly configure the IO subsystem
to make sure only one thread is ever used. With a single thread,
there are no questions about how threads are chosen etc.

Modifications
-------------
Configure the IO subsystem in this test to restrict the number
of worker threads to 1 and adjust the test accordingly.

Result
------
More robust test.

Ladicek commented:
I've finished the implementation of implicitly denying access to methods without security annotations. It no longer checks all the methods in the class, just the JAX-RS resource methods, as documented.

Ladicek commented:
Ah and in addition to fixing the `PrincipalLeakTest`, I also improved how the threads are handled. @michalszynkiewicz you might be interested: https://github.com/thorntail/thorntail/pull/1088/commits/26c76c75f6029e27cc52d165dd8cd0acf10ed10a :-)

Ladicek commented:
Windows failures are 1. the usual MP FT TCK (`BulkheadSynchRetryTest.testBulkheadClassSynchronousPassiveRetry55`), 2. not sure why but also MP Config TCK. Both seem unrelated.

kenfinnigan commented:
I'm not sufficiently versed in JWT to know whether the changes are good or not.

@sberyozkin, @michalszynkiewicz could one of you review and approve?
